sound/soc/codecs/max98373-sdw.c

```diff
@@ -0,0 +1,1203 @@
+	/* Enable ADC */
+	regmap_write(max98373->regmap, MAX98373_R2047_IV_SENSE_ADC_EN, 3);
+	/* Set default speaker volume */
+	regmap_write(max98373->regmap, MAX98373_R203D_AMP_DIG_VOL_CTRL, 0);
+	/* Enable speaker */
+	regmap_write(max98373->regmap, MAX98373_R2043_AMP_EN, 3);
+
+	/* Set default Soundwire clock */
+	regmap_write(max98373->regmap, MAX98373_R2036_SOUNDWIRE_CTRL, 5);
+	/* Set default sampling rate for speaker and IVDAC */
+	regmap_write(max98373->regmap, MAX98373_R2028_PCM_SR_SETUP_2, 0x88);
+	/* Amp enable */
+	regmap_write(max98373->regmap, MAX98373_R20FF_GLOBAL_SHDN, 1);
+
+	pm_runtime_put_sync_autosuspend(dev);
+
+	return 0;
+}
+
+
+static int max98373_snd_probe(struct snd_soc_component *component)
+{
+	struct max98373_sdw_priv *max98373 =
+		snd_soc_component_get_drvdata(component);
+
+	regmap_write(max98373->regmap, MAX98373_R20FF_GLOBAL_SHDN, 0);
+
+	/* Enable the codec - it can play the audio after 4 ms....	*/
+	regmap_write(max98373->regmap, MAX98373_R20FF_GLOBAL_SHDN, 1);
+
+	return 0;
+}
+
+static int max98373_clock_config(struct sdw_slave *slave,
+	struct sdw_bus_params *params)
+{
+	struct device *dev = &slave->dev;
+	struct max98373_sdw_priv *max98373 = dev_get_drvdata(dev);
+	unsigned int clk_freq, value;
+
+	clk_freq = (params->curr_dr_freq >> 1);
+
+	/* Before programming the clock - disable the codec */
+	regmap_write(max98373->regmap, MAX98373_R20FF_GLOBAL_SHDN, 0);
+
+	/*
+	 *	Select the proper value for the register based on the
+	 *	requested clock. If the value is not in the list,
+	 *	use reasonable default - 12 MHz
+	 */
+	switch (clk_freq) {
+	case 13000000/2:
+		value = 0x06 + 0x08;
+		break;
+	case 13000000/4:
+		value = 0x06 + 0x10;
+		break;
+	case 12288000:
+		value = 0x05;
+		break;
+	case 12288000/2:
+		value = 0x05 + 0x08;
+		break;
+	case 12288000/4:
+		value = 0x05 + 0x10;
+		break;
+	case 12288000/8:
+		value = 0x05 + 0x18;
+		break;
+	case 12000000:
+		value = 0x04;
+		break;
+	case 12000000/2:
+		value = 0x04 + 0x08;
+		break;
+	case 12000000/4:
+		value = 0x04 + 0x10;
+		break;
+	case 12000000/8:
+		value = 0x04 + 0x18;
+		break;
+	case 11289600:
+		value = 0x03;
+		break;
+	case 11289600/2:
+		value = 0x03 + 0x08;
+		break;
+	case 11289600/4:
+		value = 0x03 + 0x10;
+		break;
+	case 11289600/8:
+		value = 0x03 + 0x18;
+		break;
+	case 9600000:
+		value = 0x02;
+		break;
+	case 9600000/2:
+		value = 0x02 + 0x08;
+		break;
+	case 9600000/4:
+		value = 0x02 + 0x10;
+		break;
+	case 9600000/8:
+		value = 0x02 + 0x18;
+		break;
+	case 8400000:
+		value = 0x01;
+		break;
+	case 8400000/2:
+		value = 0x01 + 0x08;
+		break;
+	case 8400000/4:
+		value = 0x01 + 0x10;
+		break;
+	case 8400000/8:
+		value = 0x01 + 0x18;
+		break;
+	case 7680000:
+		value = 0x00;
+		break;
+	case 7680000/2:
+		value = 0x00 + 0x08;
+		break;
+	case 7680000/4:
+		value = 0x00 + 0x10;
+		break;
+	case 7680000/8:
+		value = 0x00 + 0x18;
+		break;
+
+	default:
+		value = 0x04;
```

Comment:
Is it worth to add a warning here?

Reply:
Thanks. Shall add the warning message here.